### ISSUE TITLE
ClusterClass rolloutAfter not implemented

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#5218]
The `Cluster.spec.topology.rolloutAfter` field already exists in the API, but we don't use it yet. So, I thought it's a good idea to track this via an issue.

The godoc states:
```
	// RolloutAfter performs a rollout of the entire cluster one component at a time,
	// control plane first and then machine deployments.
```

**Environment:**

- Cluster-api version: latest

/kind bug
/area topology
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by vincepri** [comment]
We need to support RolloutAfter for MachineDeployment first

**Comment by enxebre** [comment]
>We need to support RolloutAfter for MachineDeployment first
I'll get back to this https://github.com/kubernetes-sigs/cluster-api/pull/4596

**Comment by ykakarap** [comment]
/assign

Unassigned accidentally.
Will work on this after RolloutSupport for MachineDeployments is ready.

**Comment by sbueringer** [comment]
@ykakarap @fabriziopandini Would it make sense to implement this now given that MD supports rolloutAfter now?
(not sure what the semantics of the CC field would be, maybe just propagate it to KCP/MD?)

**Comment by sbueringer** [comment]
I agree that introducing a Cluster-wide rolloutAfter field without a clear idea of how it should work should be avoided. (especially given hwo non-trivial it is to implement a phased/incremental rollout (compare with what we do for version).

I wonnder though if it makes sense to expose a CP / MD -level field like we did for all the other fields. We can also just deprecate for now, but I wonder how quickly we will get a request that this field cannot be set via ClusterClass and then we need another release to provide the field

But I'm fine either way, definitely +1 to deprecating the current field as fields without any function are really bad UX.

**Comment by sbueringer** [comment]
While we're talking about deprecating fields without function. Should we consider deprecationg ProgressDeadlineSeconds on MachineDeployment? We default it, but we don't use it anywhere.

**Comment by ykakarap** [comment]
>especially given hwo non-trivial it is to implement a phased/incremental rollout (compare with what we do for version

+1. The semantics could be confusing and lead to back UX. 

>definitely +1 to deprecating the current field as fields without any function are really bad UX.

+1. 
 Can we make it part of the current release? Given that ClusterClass is behind a feature gate and also that the field is not actually used anywhere and we are only Deprecating it and not actually dropping it, it should be fine, correct? (Given our current cherry-pick rules, it does not clearly fall under any of those categories but I think we are still good to do it.)

**Comment by sbueringer** [comment]
I'm fine with doing that. As deprecating the field doesn't change anything apart from giving users more information ahead of time.

@ykakarap Can you open a PR to deprecate the rolloutAfter field and then bring it up in office hours? Basically we should discuss if we cherry-pick the PR to deprecate with v1.4 or if we have to wait until v1.5. The PR against main can be merged regardless.

**Comment by ykakarap** [comment]
I will create the PR.

**Comment by ykakarap** [comment]
Deprecation PR: https://github.com/kubernetes-sigs/cluster-api/pull/8324

**Comment by ykakarap** [comment]
/close

This issue is addressed in https://github.com/kubernetes-sigs/cluster-api/pull/8324
